Add regex_replace Spark function

codyschierbeck opened the pull request:
Add regex_replace Spark function: https://spark.apache.org/docs/latest/api/sql/#regexp_replace

Spark implementation https://github.com/apache/spark/blob/v3.2.2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L521

Spark test: https://github.com/apache/spark/blob/v3.2.2/sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala#L131

paul-amonson commented on velox/functions/lib/Re2Functions.h:
```diff
@@ -98,6 +98,14 @@ std::shared_ptr<exec::VectorFunction> makeRe2Extract(
 
 std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures();
 
+// TODO: Write comments
```
Did you miss this?

codyschierbeck replied:
Yes, new commits have comments. Thank you

paul-amonson reviewed (approved):
One non-code issue you should fix.  Other than that looks good.

akashsha1 commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -82,4 +82,18 @@ std::shared_ptr<exec::VectorFunction> makeRegexExtract(
+  // instead of just constant patterns.
+  // This was done in order to emulate specific unit tests done
+  // in Spark SQL.
+  // ensureRegexIsCompatible("REGEXP_REPLACE", inputArgs[1].constantValue);
```
consider removing comment here, and place it where you are handling it.

codyschierbeck replied:
Comment is there to explain why the function is missing compared to its neighbors. The implementation has its own comments to explain

akashsha1 commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -817,6 +1070,87 @@ std::shared_ptr<VectorFunction> makeRe2Extract(
+          .returnType("varchar")
+          .argumentType("varchar")
+          .argumentType("varchar")
+          .argumentType("varchar")
```
make varchar, integer, bigint constant string and use the constants here.

codyschierbeck replied:
Updated to leverage TypeKind enum

akashsha1 commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -817,6 +1070,87 @@ std::shared_ptr<VectorFunction> makeRe2Extract(
+      name,
+      numArgs);
+
+  VELOX_USER_CHECK(
```
loop over the three arguments and check?

codyschierbeck replied:
Doable but also not sure if necessary. Loop would need to handle the grammar of first, second, third, fourth, etc. Additionally, no other functions seem to handle it this way (with a loop). Thoughts?

yma11 commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -381,6 +406,235 @@ class Re2SearchAndExtract final : public VectorFunction {
+   * input string.
+   */
+  explicit Re2ReplaceConstantPattern(StringView basePattern)
+      : basePattern_(basePattern.getString()) {}
```
maybe rename `basePattern_` to `re_` to keep consistent with existing Re2Functions?

yma11 commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -381,6 +406,235 @@ class Re2SearchAndExtract final : public VectorFunction {
+      // Decode the position vector containing the starting position for the
+      // regex replacement.
+      exec::LocalDecodedVector positionVec(context, *args[3], rows);
+      position = positionVec->valueAt<int64_t>(0);
```
Seems we can't assume position is `int64_t` as we also support `INTEGER` in signature. Right?

codyschierbeck replied:
Removed INTEGER support. BIGINT covers.

As for moving functions to Spark, it seems like all Spark regex functions are located in velox/functions/lib. How should we handle this?

yma11 replied:
We don't need a copy for Spark if there is NO semantics mismatch between Spark and Presto.

codyschierbeck replied:
Moved implementation to just Spark as PrestoSQL allows for a function to be passed as the replacement argument which is not available for Spark

codyschierbeck commented:
> The PR structure is overall in good shape now. But seems these 4 UTs failed in the CI check, can you resolve them first?
> 
> ```
> 261: [  FAILED  ] 4 tests, listed below:
> 261: [  FAILED  ] RegexFunctionsTest.RegexReplaceRegistration
> 261: [  FAILED  ] RegexFunctionsTest.RegexReplacePosition
> 261: [  FAILED  ] RegexFunctionsTest.RegexReplaceMultipleMatches
> 261: [  FAILED  ] RegexFunctionsTest.RegexReplaceWord
> ```
> 
> I am wondering whether presto/spark have some different behaviors in this `regex_replace` function because if so, we will need move this implementation to spark functions.

Where are you seeing this? In the most recent run I am seeing all the tests pass.

codyschierbeck commented:
@assignUser I am getting fails on tests that are not changed by my PR. Specifically in velox_cache_test. Is there anything I can do to resolve this?

assignUser commented:
https://github.com/facebookincubator/velox/pull/6681 was just merged and should fix the flaky test. So a rebase should fix your issue :)

codyschierbeck commented:
All tests passed in current run after rebase. Error is now found in spark fuzzer, which is from code I have not changed

Error Source: USER
Error Code: UNSUPPORTED
Reason: Type not supported yet: VARBINARY
Retriable: False
Function: appendSqlLiteral
File: ../../velox/expression/ConstantExpr.cpp
Line: 213


Also seeing errors in uploading results

akashsha1 commented:
> All tests passed in current run after rebase. Error is now found in spark fuzzer, which is from code I have not changed
> 
> Error Source: USER Error Code: UNSUPPORTED Reason: Type not supported yet: VARBINARY Retriable: False Function: appendSqlLiteral File: ../../velox/expression/ConstantExpr.cpp Line: 213
> 
> Also seeing errors in uploading results

@assignUser - Is this something you're aware of? We have not added fuzzer tests, and are not sure if this failure is due to any missing fuzzer testing we should do.

assignUser commented:
Yeah the fuzzer runs do fail on unreleated things, that's kind of in the nature of a fuzzer so no wrires. The upload fail seems familiar but I can't recall @kgpai wasn't it related to cci quotas or something?

akashsha1 commented:
> Yeah the fuzzer runs do fail on unreleated things, that's kind of in the nature of a fuzzer so no worries. The upload fail seems familiar but I can't recall @kgpai wasn't it related to cci quotas or something?



> Yeah the fuzzer runs do fail on unreleated things, that's kind of in the nature of a fuzzer so no worries. The upload fail seems familiar but I can't recall @kgpai wasn't it related to cci quotas or something?

Would also appreciate your reviews on this change, thanks. Please help include anyone else who is familiar with this area.

mbasmanova reviewed (commented):
@codyschierbeck Would you add documentation for this newly-added function?

mbasmanova commented on velox/functions/sparksql/RegexFunctions.h:
```diff
@@ -60,4 +60,28 @@ std::shared_ptr<exec::VectorFunction> makeRegexExtract(
     const std::vector<exec::VectorFunctionArg>& inputArgs,
     const core::QueryConfig& config);
 
+std::shared_ptr<exec::VectorFunction> makeRegexReplace(
```
makeRegexReplace and makeRe2Replace function names appear the "same". What are the differences between these 2 functions? Would it be possible to come up with more descriptive names? Also, please, document makeRegexReplace function.

codyschierbeck replied:
makeRe2Replace is a forgotten function and not defined under RegexFunctions.cpp. Will remove with next commit. Only replace function is makeRegexReplace. Will document with next commit as well.

codyschierbeck replied:
Removed in recent commit

mbasmanova commented on velox/functions/lib/Re2Functions.cpp:
```diff
@@ -15,6 +15,7 @@
  */
 #include "velox/functions/lib/Re2Functions.h"
 
+#include <expression/EvalCtx.h>
```
Why is this needed?

codyschierbeck replied:
My c/c++ code insights plug-in tends to add headers it believes are needed. I removed it and ran tests which all passed.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +71,255 @@ void ensureRegexIsCompatible(
+  checkForCompatiblePattern(
+      std::string(pattern.data(), pattern.size()), functionName);
+}
+class Re2ReplaceConstantPattern final : public exec::VectorFunction {
```
Add empty line

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +71,255 @@ void ensureRegexIsCompatible(
+}
+class Re2ReplaceConstantPattern final : public exec::VectorFunction {
+ public:
+  /**
```
Please, review coding standards and update to comply.

codyschierbeck replied:
Modified class/documentation exposed comments to be triple-slashed and removed obvious comments in recent commit

mbasmanova reviewed (changes requested):
Some initial comments.

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +82,310 @@ void ensureRegexIsCompatible(
+      }
+      VELOX_CHECK(!(position - 1 < 0));
+
+      const StringView& currentOverwrite =
+          overwriteVec->valueAt<StringView>(row);
+
+      // For an unknown reason, when we do not create the StringPiece with
+      // .getString() and instead reference the StringView, the line below
+      // "strValue.substr" can overwrite the values found in overwrite. Cannot
+      // explain this behavior and must default to getString
+      std::vector<re2::StringPiece> overwrite = {
+          re2::StringPiece(currentOverwrite.getString())};
+
+      const StringView& current = input->valueAt<StringView>(row);
+      std::string strValue = current.getString();
+
+      // If position is not 1, adjust the string for regex operation.
+      // Spark implementation default position is 1. It is 1-indexed instead of
+      // 0-indexed.
+      std::string prefix;
+      if (position - 1 > 0) {
+        prefix = strValue.substr(0, position - 1);
+        if (position - 1 > strValue.length()) {
+          strValue = "";
+        } else {
+          strValue = strValue.substr(position - 1);
+        }
```
@mbasmanova I implemented the changes that you have requested. All tests passed locally, but I did include changes that allow for the constant pattern use case to be called on multiple rows at once which was left out as an oversight. 

In doing so I ran into an issue that I previously ran into with the non-constant pattern use case. The issue and relevant code are highlighted in the code referenced. Previously, overwrite was declared using just StringView, but debugging revealed that after the .substr line is run, overwrites buffer gets overwritten. This behavior does not occur when .getString() is used to create the StringPiece.

Do you have an idea as to why this is happening?

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,252 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
         fmt::format("regexp_extract(c0, '{}')", pattern), str);
   }
+
+  void regex_replace_all_params_constant(
```
Please, review coding style guidelines and update to comply.

codyschierbeck replied:
Updated. I believe they comply now

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -16,37 +16,62 @@
+// returns nullopt.
+template <typename T>
+std::optional<T> getIfConstant(const BaseVector& v) {
+  if (v.encoding() == VectorEncoding::Simple::CONSTANT &&
```
v.isConstantEncoding()

codyschierbeck replied:
Implemented

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -16,37 +16,62 @@
+template <typename T>
+std::optional<T> getIfConstant(const BaseVector& v) {
+  if (v.encoding() == VectorEncoding::Simple::CONSTANT &&
+      v.isNullAt(0) == false) {
```
!v.isNullAt(0)

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -16,37 +16,62 @@
+/**
+ * Validates the provided regex pattern to ensure its compatibility with the
+ * system. The function checks if the pattern uses features like character class
+ * union, intersection, or difference which are not supported.
```
Maybe mention these restrictions in the function documentation.

codyschierbeck replied:
These restrictions and others are mentioned in https://github.com/google/re2/wiki/Syntax link in the Spark regexp documentation

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -16,37 +16,62 @@
+ *
+ * @param pattern The regex pattern string to validate.
+ * @param functionName (Optional) Name of the calling function to include in
+ * error messages. Default is "regex_replace".
```
Do we need the default? Consider removing for readability. Is there any reason other functions do not validate the pattern?

codyschierbeck replied:
We should've used functionName instead of "Velox". Fixed in recent commit.

Right now, regex_replace is unique because it's the only spark regex function supporting non-constant patterns. So, to maintain regex compatibility, we introduced a special function that checks each pattern after they're decoded.

On the other hand, ensureRegexIsCompatible works on a VectorPtr, which is expected to have just one pattern. This function kicks in before we call RLIKE and REGEXP_EXTRACT. If we decide to allow non-constant patterns for them in the future, we could probably retire ensureRegexIsCompatible and just let them use checkForCompatiblePattern for every decoded pattern.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -16,37 +16,62 @@
             "{} does not support character class union, intersection, "
             "or difference ([a[b]], [a&&[b]], [a&&[^b]])",
-            functionName);
+            "Velox");
```
Why not just replace {} with Velox?

mbasmanova reviewed (commented):
Is there any particular reason to implement this function as vector function? Can it be simple function instead? CC: @laithsakka

mbasmanova commented:
@codyschierbeck Thank you for iterating. Would you squash all commits. GitHub becomes slow when there are too many commits.

codyschierbeck commented:
Made a mistake with rebase, please ignore

codyschierbeck commented:
@mbasmanova Hi, I am having problems with squashing these commits. Whenever I rebase and squash into one commit, GitHub no longer recognizes the commits that I merged into this branch that fixed problems with velox_memory_test, and believes that I now have 380 new commits. I am not sure how to resolve this problem.

codyschierbeck commented:
@mbasmanova Managed to resolve the issues I was having with squashing the commits. Apologies for the delay.

mbasmanova commented:
> Is there any particular reason to implement this function as vector function? Can it be simple function instead? 

@codyschierbeck Thank you for iterating. Still wondering about the above.

CC: @laithsakka

codyschierbeck commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -111,5 +359,233 @@ TEST_F(RegexFunctionsTest, RegexMatchRegistration) {
+      positions);
+}
+
+// Test to match
+// https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala#L180-L184
+// This test is the crux of why regex_replace needed to support non-constant
+// parameters. Used position {0,0} out of convenience, ideally we create another
+// function that does not pass a position parameter.
+TEST_F(RegexFunctionsTest, RegexReplaceMatchSparkSqlTest) {
+  std::vector<int64_t> positions = {1, 1, 1};
+  testingRegexReplaceRows(
+      {"100-200", "100-200", "100-200"},
+      {"(\\d+)-(\\d+)", "(\\d+)-(\\d+)", "(\\d+)"},
+      {"300", "400", "400"},
+      {{std::vector<std::string>{"300"}},
+       {std::vector<std::string>{"400"}},
+       {std::vector<std::string>{"400-400"}}},
+      positions);
+}
```
@mbasmanova I am still getting familiar with Velox as a codebase and this is my first PR so I could very easily be wrong, but from what I can tell it is a Vector function. 

These functions were modelled after the "class Re2SearchAndExtract final : public VectorFunction" and "class Re2Match final : public VectorFunction" in velox/functions/lib/Re2Functions.cpp, and I made an assumption that I would also have to create a VectorFunction. 

The constant parameters use case can be called on a with regex_replace('Hello World', 'l', 'L'), but it can also be called on a row with the same values. Additionally, if a user had a column of strings they wanted to do a batch of replacements on, they can do so with the constant pattern use case.

The non-constant parameters use case can be called with a rowVector like seen in the Spark UT test case, which attempts to directly match the Spark use case found here: https://github.com/apache/spark/blob/v3.2.2/sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala#L131 which operates on a dataframe. 

The testingRegexReplaceRows function converts the passed parameters into a rowVector, then the columns are passed into the function. 

I hope this answers your question, but to be honest I am not sure if it does. If this can be accomplished using a simple function I'd be glad to try that out, though I am still not super familiar with the everything the codebase has to offer. Another assumption I am making is that a simple function could only accomplish the regex_replace("Hello World", "l", "L") call for the constant parameters use case. Is that correct?  Or can they accomplish the same thing that the non-constant parameters use case is?

Apologies if I am misunderstanding anything or if I am being confusing. I would be glad to clarify anything.

mbasmanova replied:
@codyschierbeck Thank you for the context. This is super helpful. Consider going over https://facebookincubator.github.io/velox/develop/scalar-functions.html and playing with simple function API. It should be possible to implement this function using that. The other functions you mentioned were implemented before simple function API became as powerful as it is now.

codyschierbeck replied:
@mbasmanova Hi Maria, I am trying to get more familiar with scalar functions vs vector functions and I have a question, if you don't mind. My basic understanding right now is that if compute/storage is required between rows, OR if it produces multiple values, a function must be a vector function. If all compute/storage for a row occurs within the context of that row AND it only produces a single value then it is better suited as a simple function. Is that correct? If not, what is a good way of deducing whether a function is better represented as either option?

In this specific example, during the non-constant pattern use case if we added an optimization to store compiled regexes in a map to reduce repeated compilation work, would that prevent this function from becoming a scalar/simple function? Another such potential optimization (though I'm unclear if the extra memory is worth the compute time decrease) can be found in the constant pattern use case. If we store "visited" strings in a map with their output after replacement as the value, we can reduce repeated work in that case as we only have to search for patterns in a given string once if there are duplicates. Would this also prevent a function from becoming a scalar/simple function? Or would we be able to store these maps as properties of the struct/class and reference that with each call?

Apologies if I am way off here. Just trying to wrap my head around the differences.

mbasmanova replied:
@codyschierbeck Cody, thank you for looking into this.

All scalar functions produce only one value. Hence, regardless of whether function is implemented as simple or vector it produces single result value for each row of input.

Simple functions are stateful, e.g. you can use member variables to cache useful data to be re-used when processing different rows. See initialize() API for initializing state using constant inputs.

CC: @laithsakka 

I suggest to prototype a simple function and share that code along with any issues / questions you encounter along the way.

codyschierbeck replied:
@mbasmanova Hi Maria, yesterday I implemented the simple function version of regex_replace for SparkSQL Velox. Thank you for pointing me in the right direction and clarifying the questions I had. You were correct, it was much better suited as a simple function and the implementation was much more straightforward after a bit of ramping.

I would appreciate it if you could take a look and provide any suggestions for change. From what I can see, all of the test cases are passing and we are only seeing issues with the fuzzer. 

Thank you and please let me know if there is anything you would like me to change.

kgpai commented:
@assignUser Sorry bit out of the context - what upload failures are we talking about - is it fuzzer artifacts ? Those are best effort and if fuzzer crashes then there wont be artifacts to upload ; Like @assignUser said fuzzer can find bugs unrelated to the PR but it might still bear some investigation.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +90,82 @@ void ensureRegexIsCompatible(
+    int replacements = RE2::GlobalReplace(&stringS, patternRE, replaceSP);
+    result.resize(stringS.size());
+    std::memcpy(result.data(), stringS.data(), stringS.size());
+    return true;
```
Change this function to 'void'.

codyschierbeck replied:
Fixed

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +90,82 @@ void ensureRegexIsCompatible(
+    stringS = prefix + stringS;
+    result.resize(stringS.size());
+    std::memcpy(result.data(), stringS.data(), stringS.size());
+    return true;
```
same

codyschierbeck replied:
Fixed

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +90,82 @@ void ensureRegexIsCompatible(
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_CHECK(!(position - 1 < 0));
```
VELOX_CHECK_LT

codyschierbeck replied:
Fixed in next push

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +90,82 @@ void ensureRegexIsCompatible(
+      const arg_type<Varchar>& replace) {
+    checkForCompatiblePattern(pattern);
+    re2::RE2 patternRE(toStringPiece(pattern));
+    re2::StringPiece replaceSP = toStringPiece(replace);
```
these names do not following naming conventions adopted in this code base; please, rename

codyschierbeck replied:
Fixed in next push

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +90,82 @@ void ensureRegexIsCompatible(
+    checkForCompatiblePattern(pattern);
+    re2::RE2 patternRE(toStringPiece(pattern));
+    re2::StringPiece replaceSP = toStringPiece(replace);
+    std::string stringS = std::string(string.data(), string.size());
```
This code seems to be creating lots of intermediate strings. Is this necessary?

codyschierbeck replied:
The problem we are running into here is that the API for RE2:GlobalReplace is very rigid. It requires a std::string*, does not allow for a position parameter, and requires a StringPiece for replacement.  I made some very small changes to this code in my next push that offer slight optimizations (less copying and more StringPiece leveraging), but to the extent of my knowledge, there is no way to avoid intermediate strings with our specific use case.

mbasmanova reviewed (commented):
@codyschierbeck Cody, I'm happy to hear you were able to re-write this function using simple API. I appreciate this is quite non-trivial to figure out (especially for the first). Great job. I'll take a closer look at the PR next week, but here are some quick comments. Have a nice weekend.

CC: @laithsakka

codyschierbeck commented:
@mbasmanova Hello, in my latest commit I implemented the changes you requested, leveraged StringPiece to reduce the number of copies when dealing with position parameter string slicing, and added a regex pattern cache which reduced the RegexFunctionsTest.RegexReplaceMassiveVectors test runtime from ~3322ms to ~870ms on average. 

This test case creates a RowVector that matches the 3-row Spark Unit test but concatenates it 100,000 times. The cache reduces 300,000 regex compilations to 2 (there is a repeated regex in the test case originally). Unfortunately, this cache does not resolve the problems/concerns you raised in #6804, especially if there aren't repeated regular expressions, but can be helpful in some corner cases. 

If we do come up with a solution for many unrepeated regexes, I would be glad to implement them into this function.

mbasmanova commented:
@codyschierbeck I haven't looked at the latest code yet, but would you clarify whether there is a cap on number of regex compilations per function instance. If there is no such cap, let's introduce it.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+  }
+
+ private:
+  mutable std::unordered_map<std::string, std::unique_ptr<re2::RE2>>
```
- use F14Map
- move member variables after methods

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+  mutable std::unordered_map<std::string, std::unique_ptr<re2::RE2>>
+      patternCache_;
+
+  re2::RE2* getCachedRegex_(const std::string& pattern) const {
```
naming: getCachedRegex_ -> getCachedRegex (no trailing underscore)

codyschierbeck replied:
Implemented

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+
+    auto patternRegex = std::make_unique<re2::RE2>(pattern);
+    auto* rawPatternRegex = patternRegex.get();
+    patternCache_[pattern] = std::move(patternRegex);
```
Let's make sure there is a hard cap on the total number of regexes.

codyschierbeck replied:
@mbasmanova Okay, I can do that. Is there a number you have in mind? I can leverage the size of the cache to check, and then throw an exception if it exceeds the limit. Off the top of my head, I can't think of a way to throw that exception while also finishing any processing for future repeated regexes, but I am not sure that matters.

mbasmanova replied:
> Is there a number you have in mind? 

I don't have a good reference, but maybe we can start with 20.

codyschierbeck replied:
Okay, I implemented a kMaxCompiledRegexes in velox/functions/lib/Re2Functions.h so any regex function can leverage it at a later date. I also added two test cases. One that checks if a VeloxRuntimeError is thrown when we reach the limit. This one runs successfully. And I added another that repeats that use case with slight changes. 

Instead of passing the limit, it hits the limit and then concatenates the RowVector 50,000 times, so we do a total of 20 * 50000 = 1,000,000 replacements but never reach the limit. This runs in 2197ms on my machine and exists successfully.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+    if (it != patternCache_.end()) {
+      return it->second.get();
+    }
+    VELOX_CHECK_LT(patternCache_.size(), kMaxCompiledRegexes);
```
Would you make this a user check and add a clear error message?

mbasmanova replied:
Let's also document this limitation.

codyschierbeck replied:
@mbasmanova Implementing all of your change requests today. Will update when complete

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+    VELOX_CHECK_LT(patternCache_.size(), kMaxCompiledRegexes);
+    auto patternRegex = std::make_unique<re2::RE2>(pattern);
+    auto* rawPatternRegex = patternRegex.get();
+    patternCache_[pattern] = std::move(patternRegex);
```
@laithsakka Implementing your changes today, will update when complete

codyschierbeck replied:
Used emplace function

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -14,31 +14,64 @@
+// If v is a non-null constant vector, returns the constant value. Otherwise
+// returns nullopt.
+template <typename T>
+std::optional<T> getIfConstant(const BaseVector& v) {
```
Removed function

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -13,32 +13,66 @@
+#include <re2/stringpiece.h>
 #include "velox/functions/lib/Re2Functions.h"
+#include "velox/functions/lib/RegistrationHelpers.h"
 
```
Removed newline and ordered includes

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -13,32 +13,66 @@
+  return std::nullopt;
+}
+
+void checkForBadPattern(const RE2& re) {
```
Added a call to this function after a new regex is compiled

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+// Blocks patterns that contain character class union, intersection, or
+// difference because these are not understood by RE2 and will be parsed as a
+// different pattern than in java.util.regex.
+void ensureRegexIsCompatible(
```
changed name

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+struct RegexReplaceFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
```
Removed FOLLY_ALWAYS_INLINE

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+    RE2::GlobalReplace(&string, *patternRegex, replaceStringPiece);
+
+    // Prepare the final result string without unnecessary copying
+    result.resize(string.size());
```
Guarded

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -13,32 +13,66 @@
   const char* charClassStart = nullptr;
   // This minimal regex parser looks just for the class begin/end markers.
-  for (const char* c = pattern.begin(); c < pattern.end(); ++c) {
+  for (const char* c = pattern.data(); c < pattern.data() + pattern.size();
```
I was getting errors with operators not being supported between const char* and std::__cxx11::basic_string<char>::const_iterator. I am unsure why this error didn't appear before, but to fix it I switched to .data()

mbasmanova replied:
@codyschierbeck Are you saying you were getting errors without any code changes. If that's the case, there might be some issue with your local setup. Let's dig in to figure it out.

codyschierbeck commented:
@mbasmanova @laithsakka I implemented most of the changes requested. Kept a single change as it resolved a bug I was running into, explained in a comment below @laithsakka question about it

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,44 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+.. spark:function:: regex_replace(string, pattern, overwrite) -> varchar
+
+    Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite``. If no match is found, the original string is returned as is.
+    There is a limit to the number of unique regexes to be compiled per function call (kMaxCompiledRegexes defined in velox/functions/lib/Re2Functions.h).
```
It would be inconvenient for the reader to search the code for kMaxCompiledRegexes. Let's hard-code the number here.

codyschierbeck replied:
Fixed

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,44 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+
+    Parameters:
+
+    - **string**: The string to be searched.
```
Would you generate docs locally, verify the formatting and share the screenshot?

codyschierbeck replied:
<img width="939" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/572ae4f4-a155-4e8f-a5c3-3f2c51ecdf51">

codyschierbeck replied:
Added the screenshot of docs. Please let me know if this is as expected

akashsha1 replied:
one minor comment here - please delete line56 
:noindex: 

as it results in unnecessary tabbing of the regex replace section

mbasmanova replied:
@codyschierbeck Looks like there is an extra indentation for the description of the second function.

codyschierbeck replied:
Looks like the :noindex: line was causing that indent difference. Removed it and now it looks as follows.

![MicrosoftTeams-image](https://github.com/facebookincubator/velox/assets/14911410/d62b8dd2-a9f4-4203-9321-56c4b65943b9)

codyschierbeck replied:
I did some digging and found the original issue. The :noindex: line had one less space than the rest of the lines that proceeded it, pushing them back an entire indent. The problem was fixed by adding that additional space. Squashed this change and the incorrect change together with the general doc changes.

codyschierbeck replied:
@mbasmanova Below is a screenshot of the docs after adding the single space

<img width="949" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/473f81d0-5621-45a1-9495-b0898d1cda6f">

mbasmanova replied:
Looks nice. Thank you.

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,44 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+
+    ::
+
+        SELECT regex_replace('Hello, World!', 'l', 'L'); -- 'HeLLo, WorLd!'
```
Would it be possible to include examples with regular expressions? This example uses a simple string.

codyschierbeck replied:
Fixed

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,44 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+
+    Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite`` starting from the specified ``position``. If the ``position`` is less than one, the function returns the original ``string`` without any modifications. If ``position`` is greater than the length of ``string``, the function returns an error.
+
+    This function is 1-indexed, meaning the first character in ``string`` is considered position 1.
```
perhaps, "Positions of the first character is 1."

codyschierbeck replied:
Fixed

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,44 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+    - **string**: The string to be searched.
+    - **pattern**: The regular expression pattern that is searched for in the string.
+    - **overwrite**: The string that replaces the substrings in ``string`` that match the ``pattern``.
+    - **position**: The 1-based index position in ``string`` to start the search.
```
typo: "index position"

perhaps, The position to start from in terms of number of characters. 1 means to start from the beginning of the string. 3 means to start from the 3-rd character.

codyschierbeck replied:
Fixed

mbasmanova commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,44 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+.. spark:function:: regex_replace(string, pattern, overwrite, position) -> varchar
+   :noindex:
+
+    Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite`` starting from the specified ``position``. If the ``position`` is less than one, the function returns the original ``string`` without any modifications. If ``position`` is greater than the length of ``string``, the function returns an error.
```
Please, mention the limit on the number of regular expressions.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.h:
```diff
@@ -60,4 +60,18 @@ std::shared_ptr<exec::VectorFunction> makeRegexExtract(
+/// regex_replace(string, pattern, overwrite, position) → string
+///
+/// If a string has a substring that matches the given pattern, replace
+/// the match in the string wither overwrite and return the string. If
```
typo: "wither"

mbasmanova commented on velox/functions/sparksql/RegexFunctions.h:
```diff
@@ -60,4 +60,18 @@ std::shared_ptr<exec::VectorFunction> makeRegexExtract(
+///
+/// If a string has a substring that matches the given pattern, replace
+/// the match in the string wither overwrite and return the string. If
+/// optional paramter position is provided, only make replacements
```
typo: "paramter"

mbasmanova commented on velox/functions/sparksql/RegexFunctions.h:
```diff
@@ -60,4 +60,18 @@ std::shared_ptr<exec::VectorFunction> makeRegexExtract(
+/// optional paramter position is provided, only make replacements
+/// after that positon in the string (1 indexed).
+///
+/// If position <= 0, throw error.
```
Can you document this? Currently, the documentation doesn't seem to clarify this.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,119 @@ void ensureRegexIsCompatible(
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace) {
+    checkForCompatiblePattern(pattern);
```
Can we first see if we have this pattern in the cache and if not, only then run this check, i.e. move this check into getCachedRegex?

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,119 @@ void ensureRegexIsCompatible(
+    std::string string(stringInput.data(), stringInput.size());
+    RE2::GlobalReplace(&string, *patternRegex, replaceStringPiece);
+
+    // Prepare the final result string without unnecessary copying
```
This comment is strange since it is followed by a memcpy. Is this comment needed? What are you trying to clarify?

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,119 @@ void ensureRegexIsCompatible(
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_CHECK_LT(-1, position - 1);
+    checkForCompatiblePattern(pattern);
```
same

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,119 @@ void ensureRegexIsCompatible(
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+    re2::StringPiece inputStringPiece(stringInput.data(), stringInput.size());
+
+    const char* prefixEnd = inputStringPiece.data() +
```
I assume position is in terms of characters, not bytes. Is this correct? Would you add some tests with non-ASCII strings?

mbasmanova replied:
I still wonder about this.

codyschierbeck replied:
The test regexReplaceNonAsciiPosition tests the following function call

regex_replace("♫ Resume is updated¡♫ ", "¡", "!", 2) and expects the output to be "♫ Resume is updated!♫ ". When compiled and run this test passes

mbasmanova replied:
@codyschierbeck I'm not sure it is sufficient to say that a test passes. If 'position' indicates number of characters, then this code seems incorrect. I might be wrong though. Please, help explain how this code works.

mbasmanova replied:
@codyschierbeck I didn't see "// Adjust the position for UTF-8 by counting the code points." code earlier. I see it now and it makes sense. Maybe use existing helper functions to skip n UTF-8 characters.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,119 @@ void ensureRegexIsCompatible(
+    VELOX_USER_CHECK_LT(
+        patternCache_.size(),
+        kMaxCompiledRegexes,
+        "regex_replace hit the maximum number of unique regexes: kMaxCompiledRegexes");
```
The user would not know what kMaxCompiledRegexes is. Please, use the actual value.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -111,5 +330,275 @@ TEST_F(RegexFunctionsTest, RegexMatchRegistration) {
   EXPECT_THROW(regexp_extract("[]", "[a[b]]"), VeloxUserError);
 }
 
+TEST_F(RegexFunctionsTest, RegexReplaceRegistration) {
```
naming: test names must start with lowercase letter

codyschierbeck replied:
Renamed all rlike, regexp_extract, and regex_replace tests to lowercase.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+
+  void testRegexReplaceSimpleOnce(
+      const std::vector<std::optional<std::string>>& input,
+      std::string pattern,
```
pass strings as const &

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+      std::string replace,
+      std::string output,
+      std::optional<int> position =
+          std::nullopt /*adding the optional position parameter*/) {
```
> /*adding the optional position parameter*/

Redundant comment. Let's remove.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+      std::string pattern,
+      std::string replace,
+      std::string output,
+      std::optional<int> position =
```
int -> int32_t

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+    };
+    auto result = [&] {
+      auto inputString =
+          makeFlatVector<StringView>(input.size(), valueAt, nullAt);
```
Can you use `makeFlatVector<std::string>(input)`?

mbasmanova replied:
Still wonder about this.

codyschierbeck replied:
makeFlatVector<std::string>(input); results in a no matching function call error, but using makeFlatVector<std::string>(input.size(), valueAt, nullAt); compiles and runs with all tests passing.

I changed all cases of makeFlatVector<StringView> to std::string using the second parameter list.

mbasmanova replied:
makeFlatVector<std::string>(input) or similar should work. Please, take a deeper look. Current code is too verbose.

codyschierbeck replied:
By converting all parameters (except position) to non-optional, the std::string(input) function call worked. This allowed me to remove all valueAt and nullAt lines.

I had to keep position as an optional parameter as it truly is an optional parameter, unlike the others.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+                "regex_replace(c0, '{}', '{}', {})",
+                pattern,
+                replace,
+                *position),
```
position.value()

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+      }
+    }();
+
+    EXPECT_EQ(result, output);
```
It would be better to move this check to the caller. Have this function return result and have the caller call EXPECT_EQ. This way the code will be more readable and the failure will point to the line in the test that failed.

codyschierbeck replied:
Fixed with new method of testing in latest commit.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+      const std::vector<std::optional<std::string>>& pattern,
+      const std::vector<std::optional<std::string>>& replace,
+      const std::vector<std::optional<std::string>>& output,
+      const std::optional<std::vector<int64_t>>& positionOpt = std::nullopt,
```
positionOpt -> position

codyschierbeck replied:
Fixed

mbasmanova replied:
Next time, consider not to reply "Fixed" on each and every comment. This generates an e-mail per reply making it noisy. Consider replying only if you have something substantial to say. For the rest, just comment one saying everything is addressed or replied to. Thanks.

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+      const std::vector<std::optional<std::string>>& output,
+      const std::optional<std::vector<int64_t>>& positionOpt = std::nullopt,
+      int repeatCount = 1) {
+    if (repeatCount < 1) {
```
Better to assert that repeatCount >= 1. Otherwise, folks may write tests that do nothing without realizing that.

codyschierbeck replied:
Fixed with EXPECT_GT(repeatCount, 0) in new commit

codyschierbeck replied:
Had to make EXPECT because ASSERT caused compile error since it does not match function return statement

mbasmanova commented on velox/functions/sparksql/tests/RegexFunctionsTest.cpp:
```diff
@@ -41,6 +43,223 @@ class RegexFunctionsTest : public test::SparkFunctionBaseTest {
+      return;
+    }
+
+    // Repeat the inputs
```
Why are you doing that? BTW, please, make sure all comments are full sentences and end with a period.

codyschierbeck replied:
Repeating input allows me to test dataframes with hundreds of thousands if not millions of rows easily. These DF's still have 20 maximum unique regexes

mbasmanova reviewed (commented):
Some comments.

codyschierbeck commented:
@mbasmanova Apologies, I was out for the past week. I have implemented the changes you requested as far as I can tell. Please let me know if I missed anything.

codyschierbeck commented:
@mbasmanova Is this pull request ready for sign-off or are there other changes you suggest?

mbasmanova commented:
@codyschierbeck I'm running behind on PRs. Will try to catch up tomorrow, but, please, ping me on Thu in case you don't hear from me.

CI is red. Please, take a look.

codyschierbeck commented:
> @codyschierbeck I'm running behind on PRs. Will try to catch up tomorrow, but, please, ping me on Thu in case you don't hear from me.
> 
> CI is red. Please, take a look.

@mbasmanova Hi, I fixed the formatting issues that were causing the CI failures. The only remaining ones are the fuzzer (discussed previously) and uploading results. I did notice that I missed some review comments so I force-pushed those changes, including the additional non-ascii test.

mbasmanova commented:
@codyschierbeck Looks like fuzzer failure is related to this PR. Please, take a look.

```
E20231026 18:35:24.263033 198765 Exceptions.h:69] Line: ../../velox/functions/sparksql/RegexFunctions.cpp:145, Function:call, Expression: -1 < position - 1 (-1 vs. -5432941019078731721), Source: RUNTIME, ErrorCode: INVALID_STATE
E20231026 18:35:24.264618 198765 ExpressionVerifier.cpp:176] Common eval: Exceptions other than VeloxUserError are not allowed.
```

codyschierbeck commented:
hi @mbasmanova @laithsakka, went through and added changes that I missed, and rebased on recent changes to fix the fuzzer issues I was running into. Please take a look at let me know if you think this PR is ready.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,116 @@ void ensureRegexIsCompatible(
+    std::string string(stringInput.data(), stringInput.size());
+    RE2::GlobalReplace(&string, *patternRegex, replaceStringPiece);
+
+    if (string.size()) {
```
What if string is empty? Should we make sure to set result to empty string as well?

codyschierbeck replied:
The default value for out_type<Varchar> is "", but I put an else statement in both call functions that resizes result to 0 to ensure it equals "".

codyschierbeck replied:
This change is now visible in the code, please let me know if it is correct

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,116 @@ void ensureRegexIsCompatible(
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_USER_CHECK_LT(-1, position - 1);
```
This check is confusing. Why not VELOX_USER_CHECK_GE(position, 1)?

Please, add a message.

codyschierbeck replied:
Used CHECK_GE and added a message

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,116 @@ void ensureRegexIsCompatible(
+
+    re2::RE2* patternRegex = getCachedRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+    re2::StringPiece inputStringPiece(stringInput.data(), stringInput.size());
```
why not toStringPiece?

codyschierbeck replied:
Used toStringPiece

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -69,7 +203,7 @@ std::shared_ptr<exec::VectorFunction> makeRLike(
   // Return any errors from re2Search() first.
   auto result = makeRe2Search(name, inputArgs, config);
-  ensureRegexIsCompatible("RLIKE", inputArgs[1].constantValue);
+  ensureRegexIsConstantAndCompatible("RLIKE", inputArgs[1].constantValue);
```
Function name here is all caps but somewhere else I saw it all lower case. Should it be lower case everywhere?

codyschierbeck replied:
I made changes and made all function names upper case to match other functions found in the file. Not sure what the standard is but changed for consistency.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.h:
```diff
@@ -60,4 +60,18 @@ std::shared_ptr<exec::VectorFunction> makeRegexExtract(
     const core::QueryConfig& config);
 
+/// Full implementation of RegexReplace found in SparkSQL only,
+/// due to semantic mismatches betweeen Spark and Presto
```
typos or incomplete sentence; also this comment seems to be duplicated; let's not duplicate comments like this to avoid diverging

codyschierbeck replied:
Removed

mbasmanova reviewed (commented):
Some comments. Looks like earlier comments / questions still are not addressed.

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,116 @@ void ensureRegexIsCompatible(
+    const char* functionName,
+    const VectorPtr& patternVector) {
+  if (!patternVector ||
+      patternVector->encoding() != VectorEncoding::Simple::CONSTANT) {
```
@mbasmanova Implementing this change causes test cases for REGEX_EXTRACT and RLIKE to fail. 

<img width="313" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/208d5659-ff3b-4838-b8c2-fe6ab163a0b5">

Below is a picture of 3 test cases, the first and the third are the failing cases.  For AllowSimpleConstantRegex the first EXPECT_EQ fails. For RegexMatchRegistration the only EXPECT_EQ fail. I'm not sure why this is occurring

<img width="256" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/fc7cf4ba-4ae3-4888-9a16-e3bf6e372255">

mbasmanova replied:
This is very strange. Something is going on. Please, take a deeper look.

```
  bool isConstantEncoding() const {
    return encoding_ == VectorEncoding::Simple::CONSTANT;
  }
```

codyschierbeck replied:
Resolved issue, isConstantEncoding working after fixing typo

codyschierbeck commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,46 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+.. spark:function:: regex_replace(string, pattern, overwrite) -> varchar
+
+    Replaces all substrings in ``string`` that match the regular expression ``pattern`` with the string ``overwrite``. If no match is found, the original string is returned as is.
+    There is a limit to the number of unique regexes to be compiled per function call (kMaxCompiledRegexes defined in velox/functions/lib/Re2Functions.h and is set to 20).
```
Fixed

codyschierbeck commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,46 @@ See https://github.com/google/re2/wiki/Syntax for more information.
+    - **string**: The string to be searched.
+    - **pattern**: The regular expression pattern that is searched for in the string.
+    - **overwrite**: The string that replaces the substrings in ``string`` that match the ``pattern``.
+    - **position**: The position to start from in terms of number of characters. 1 means to start from the beginning of the string. 3 means to start from the 3rd character. Positions less than or equal to zero throw an error.
```
This was correct, the earlier statement was wrong. Fixed

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -13,32 +13,56 @@
-      patternVector->as<ConstantVector<StringView>>()->valueAt(0);
+}
+
+/// Validates the provided regex pattern to ensure its compatibility with the
```
Changed all .cpp comments to //

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -13,32 +13,56 @@
+/// class union, intersection, or difference which are not supported in C++ RE2
+/// library but are supported in Java regex.
+///
+/// This function should be called on the indivdual patterns of a decoded
```
Fixed

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -13,32 +13,56 @@
+/// other functions to enable non-constant patterns in the future.
+void checkForCompatiblePattern(
+    const std::string& pattern,
+    const char* functionName = "regex_replace") {
```
Removed default value

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,116 @@ void ensureRegexIsCompatible(
+  }
+  const StringView pattern =
+      patternVector->as<ConstantVector<StringView>>()->valueAt(0);
+  // Call the factored out function to check the pattern.
```
Removed

codyschierbeck commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,116 @@ void ensureRegexIsCompatible(
+}
+
+/// Full implementation of RegexReplace found in SparkSQL only,
+/// due to semantic mismatches betweeen Spark and Presto
```
Removed

codyschierbeck commented:
@mbasmanova @laithsakka I made some changes to the testing functions to make the code less verbose. I also implemented the requested changes. Please take a look and let me know what you think.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,128 @@ void ensureRegexIsCompatible(
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+    re2::StringPiece inputStringPiece = toStringPiece(stringInput);
+
+    // Adjust the position for UTF-8 by counting the code points.
```
I wonder if we could use an existing helper function.

```
/// Return the size in bytes for the char pointed to by u_input.
/// This function is not part of the original utf8proc, it is a simplified
/// verion of utf8proc_codepoint. It assumes a valid utf8 input otherwise output
/// is undefined.
UTF8PROC_DLLEXPORT int utf8proc_char_length(const char* u_input) {
```

codyschierbeck replied:
I will look into this and try to implement tomorrow morning if I can make it work

codyschierbeck replied:
I implemented a version that uses utf8proc_char_length along with a new test case for when a character is not recognized and returns -1.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,128 @@ void ensureRegexIsCompatible(
+      ++utf8Position;
+    }
+
+    if (utf8Position >= stringInput.size()) {
```
In this case, we do not need the following logic including creation of RE2::GlobalReplace. We can just set result to stringInput, no?

codyschierbeck replied:
Correct, I will implement this change tomorrow morning.

codyschierbeck replied:
@mbasmanova I implemented this logic but had to change the if statement clause to "utf8Position > stringInput.size()" to handle the case covered in regexReplaceEmptyStringsAndPatterns where we run regex_replace("", "", "prefix") -> "prefix".

Please let me know if you see anything else you would like me to change.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,135 @@ void ensureRegexIsCompatible(
+
+    // Adjust the position for UTF-8 by counting the code points.
+    size_t utf8Position = 0;
+    size_t codePointsCounted = 0;
```
nit: numCodePoints

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,135 @@ void ensureRegexIsCompatible(
+    size_t utf8Position = 0;
+    size_t codePointsCounted = 0;
+    while (codePointsCounted < position - 1 &&
+           utf8Position < stringInput.size()) {
```
should this be <= ?

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,135 @@ void ensureRegexIsCompatible(
+           utf8Position < stringInput.size()) {
+      int charLength =
+          utf8proc_char_length(inputStringPiece.data() + utf8Position);
+      if (charLength <= 0) {
```
VELOX_USER_CHECK_GT(charLength, 0, "regex_replace encountered invalid UTF-8 character")

Use lowercase for function names. Do not include invalid characters in the error message. These won't come out nice anyway.

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -57,6 +81,135 @@ void ensureRegexIsCompatible(
+      ++codePointsCounted;
+      utf8Position += charLength;
+    }
+    if (utf8Position > stringInput.size()) {
```
>= ?

codyschierbeck replied:
This was changed to just greater than to cover when we have both an input string and a pattern that are empty, but not the replace value.

mbasmanova replied:
I might be missing something, but wouldn't be it clearer to check for empty input and pattern at the start of this function?

mbasmanova reviewed (approved):
@codyschierbeck Thank you for iterating on this PR. Looks great % a couple of nits.

FelixYBW commented:
@rui-mo

mbasmanova commented:
@codyschierbeck CI is red. Would you take a look?

codyschierbeck commented:
> @codyschierbeck CI is red. Would you take a look?

Hi @mbasmanova, I took a look at the fuzzer in linux-build and I don't understand why it is failing. The fuzzer terminates after RegexFunctions.cpp line 153 which is the VELOX_USER_CHECK_GT(charLength, 0, "regex_replace encountered invalid UTF-8 character"); line. This error is expected to be thrown.

I tried to replicate this locally but the fuzzer uses ${RANDOM} for seed. I looked through the rest of the fuzzer and it seemed fine but I still don't understand why it terminated after encountering an expected error.

In linux-pr-fuzzer-run I checked each VeloxUserError occurance and none of them referred to RegexFunctions.cpp, so I do not know why it failed there. It may have occured below line 103121 but that's as far as the big view would let me see.

I talked with @assignUser on Slack and let him know the issues I was running into.

mbasmanova commented:
@codyschierbeck Have you tried reproducing the failures using seed? See https://facebookincubator.github.io/velox/develop/testing/fuzzer.html#how-to-reproduce-failures

mbasmanova commented:
The error is:

```
I20231108 19:23:11.234970 224813 ExpressionFuzzer.cpp:1313] ==============================> Started iteration 150 (seed: 964558807)
I20231108 19:23:11.236704 224813 ExpressionVerifier.cpp:86] Executing expression 0 : upper(left((sha1("c0"),;~,"c1",doy(539141144)),2100894719))
I20231108 19:23:11.239480 224813 FuzzerToolkit.cpp:103] Exceptions match.
I20231108 19:23:11.239521 224813 ExpressionFuzzer.cpp:1353] Both paths failed with compatible exceptions. Retrying expression using try().
I20231108 19:23:11.239557 224813 ExpressionVerifier.cpp:86] Executing expression 0 : try(upper(left((sha1("c0"),;~,"c1",doy(539141144)),2100894719)))
I20231108 19:23:11.241263 224813 ExpressionVerifier.cpp:74] All results match.
I20231108 19:23:11.242923 224813 ExpressionVerifier.cpp:74] All results match.
I20231108 19:23:11.243070 224813 ExpressionVerifier.cpp:74] All results match.
I20231108 19:23:11.243163 224813 ExpressionFuzzer.cpp:1274] Retrying original expression on 78 rows without errors
I20231108 19:23:11.243212 224813 ExpressionVerifier.cpp:86] Executing expression 0 : upper(left((sha1("c0"),;~,"c1",doy(539141144)),2100894719))
I20231108 19:23:11.244624 224813 ExpressionVerifier.cpp:74] All results match.
E20231108 19:23:11.245645 224813 FuzzerToolkit.cpp:95] Only simplified path threw exception:
I20231108 19:23:11.246937 224813 ExpressionVerifier.cpp:346] Persisted input: --fuzzer_repro_path /tmp/spark_fuzzer_repro/velox_expressionVerifier_MMxTym --input_path /tmp/spark_fuzzer_repro/velox_expressionVerifier_MMxTym/input_vector --result_path /tmp/spark_fuzzer_repro/velox_expressionVerifier_MMxTym/result_vector --sql_path /tmp/spark_fuzzer_repro/velox_expressionVerifier_MMxTym/sql
terminate called after throwing an instance of 'facebook::velox::VeloxUserError'
  what():  Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: (-1 vs. 0) regex_replace encountered invalid UTF-8 character
Retriable: False
Expression: charLength > 0
Function: call
File: ../../velox/functions/sparksql/RegexFunctions.cpp
Line: 153
```

codyschierbeck commented:
> The error is:
> 
> ```
> I20231108 19:23:11.234970 224813 ExpressionFuzzer.cpp:1313] ==============================> Started iteration 150 (seed: 964558807)
> I20231108 19:23:11.236704 224813 ExpressionVerifier.cpp:86] Executing expression 0 : upper(left((sha1("c0"),;~,"c1",doy(539141144)),2100894719))
> I20231108 19:23:11.239480 224813 FuzzerToolkit.cpp:103] Exceptions match.
> I20231108 19:23:11.239521 224813 ExpressionFuzzer.cpp:1353] Both paths failed with compatible exceptions. Retrying expression using try().
> I20231108 19:23:11.239557 224813 ExpressionVerifier.cpp:86] Executing expression 0 : try(upper(left((sha1("c0"),;~,"c1",doy(539141144)),2100894719)))
> I20231108 19:23:11.241263 224813 ExpressionVerifier.cpp:74] All results match.
> I20231108 19:23:11.242923 224813 ExpressionVerifier.cpp:74] All results match.
> I20231108 19:23:11.243070 224813 ExpressionVerifier.cpp:74] All results match.
> I20231108 19:23:11.243163 224813 ExpressionFuzzer.cpp:1274] Retrying original expression on 78 rows without errors
> I20231108 19:23:11.243212 224813 ExpressionVerifier.cpp:86] Executing expression 0 : upper(left((sha1("c0"),;~,"c1",doy(539141144)),2100894719))
> I20231108 19:23:11.244624 224813 ExpressionVerifier.cpp:74] All results match.
> E20231108 19:23:11.245645 224813 FuzzerToolkit.cpp:95] Only simplified path threw exception:
> I20231108 19:23:11.246937 224813 ExpressionVerifier.cpp:346] Persisted input: --fuzzer_repro_path /tmp/spark_fuzzer_repro/velox_expressionVerifier_MMxTym --input_path /tmp/spark_fuzzer_repro/velox_expressionVerifier_MMxTym/input_vector --result_path /tmp/spark_fuzzer_repro/velox_expressionVerifier_MMxTym/result_vector --sql_path /tmp/spark_fuzzer_repro/velox_expressionVerifier_MMxTym/sql
> terminate called after throwing an instance of 'facebook::velox::VeloxUserError'
>   what():  Exception: VeloxUserError
> Error Source: USER
> Error Code: INVALID_ARGUMENT
> Reason: (-1 vs. 0) regex_replace encountered invalid UTF-8 character
> Retriable: False
> Expression: charLength > 0
> Function: call
> File: ../../velox/functions/sparksql/RegexFunctions.cpp
> Line: 153
> ```

@mbasmanova  I used that seed and was able to reproduce both with the original VELOX_USER_FAIL line and the VELOX_USER_CHECK_GT. 

I changed the VELOX_USER_CHECK_GT to VELOX_CHECK_GT and got more information from the stack, specifically the context and top-level context

terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (-1 vs. 0) regex_replace encountered invalid UTF-8 character
Retriable: False
Expression: charLength > 0
Context: (sha1(c0), ;~:VARCHAR, c1, 116:BIGINT)
Top-Level Context: upper(left((sha1(c0), ;~:VARCHAR, c1, 116:BIGINT), 2100894719:INTEGER))
Function: call
File: /home/sdp/pinterest/velox/velox/functions/sparksql/RegexFunctions.cpp
Line: 153

Neither of the contexts appear to be calling regex_replace.

I found SparkExpressionFuzzerTest.cpp in the stack trace: (_build/debug/velox/expression/tests/spark_expression_fuzzer_test(_start+0x24)[0x55e20cfbaec4])

In that file there appears to be list of skipFunctions to not run the fuzzer on. I added regex_replace to the list of skipFunctions in SparkExpressionFuzzerTest.cpp and re-ran the fuzzer with the given seed, and it still failed.

I will take another look tomorrow morning.

codyschierbeck commented:
@assignUser Do you know why the fuzzer is calling regex_replace is in upper(left((sha1(c0), ;:VARCHAR, c1, 116:BIGINT), 2100894719:INTEGER))? My best guess as to why the fuzzer is failing is because it expects an output from the above expression but at some point (sha1(c0), ;:VARCHAR, c1, 116:BIGINT) calls regex_replace with an invalid input causing regex_replace to throw an exception that the fuzzer does not expect. 

I looked at the sha1 spark implementation and I can't find any reason why it would call regex_replace

mbasmanova commented on velox/functions/sparksql/RegexFunctions.cpp:
```diff
@@ -78,8 +227,21 @@ std::shared_ptr<exec::VectorFunction> makeRegexExtract(
+      Varchar,
+      Varchar,
+      Varchar,
+      int64_t>({prefix, "REGEX_REPLACE"});
```
@codyschierbeck There is a bug here which causes this function to be registered with 2 names: prefix and "REGEX_REPLACE". The fuzzer uses empty prefix which causes the function to be registered with empty name. This then makes it confusing that an expression doesn't seem to have regex_replace function when in fact it does.

To fix, replace prefix, with prefix +.

We should also add a check to function registry to not allow registering functions with empty names. CC: @laithsakka @kgpai @bikramSingh91 

Would you help make this change?

codyschierbeck commented:
Hi @mbasmanova, could you help merge this PR? All validation is passing now. 

For the issue you mentioned about empty function aliases, I have a draft PR located here #7569

codyschierbeck commented:
@mbasmanova What are the next steps to merge this?

mbasmanova commented:
@codyschierbeck Last time I looked there were Fuzzer failures. Are these fixed now? If so, please, rebase and I'll try to merge.

codyschierbeck commented:
@mbasmanova Were those the fuzzer failures that came from registering a function without a name? I resolved that issue with your recommended solution and then handled some merge conflicts we were seeing with [8e86d7e](https://github.com/facebookincubator/velox/pull/6516/commits/8e86d7ecd5df3b34eae815b931f42d3fe391736a). The fuzzer doesn't appear to be failing right now, but I may be missing something.

mbasmanova commented:
> Were those the fuzzer failures that came from registering a function without a name? 

Yes. I was referring to these. If there are fixed now, then, please, rebase and I'll try to merge.

codyschierbeck commented:
@mbasmanova Seeing a Facebook Internal-Linter issue. It says there is 1 advice and 1 warning. but I can't see details besides that.

mbasmanova commented:
@codyschierbeck A few linter warnings:

<img width="489" alt="Screenshot 2023-11-27 at 4 17 16 PM" src="https://github.com/facebookincubator/velox/assets/27965151/c041a48f-4ec0-4073-acef-62392709562f">
<img width="488" alt="Screenshot 2023-11-27 at 4 17 29 PM" src="https://github.com/facebookincubator/velox/assets/27965151/0c34e13a-d087-42fb-a7d5-f875e4b50fac">
<img width="488" alt="Screenshot 2023-11-27 at 4 17 40 PM" src="https://github.com/facebookincubator/velox/assets/27965151/2c871fdc-58e0-4306-9554-ebee8a4f3998">

codyschierbeck commented:
@mbasmanova Hi, is there a way for me to run this Linter on my end? I saw it failed again so I tested each import and found more unused. I don't want you to have to reimport and grab screenshots for each time it fails. 

I have those changes ready, but I don't want to push in case there's something in this Linter run I missed.

mbasmanova commented:
@codyschierbeck 

> is there a way for me to run this Linter on my end?

This is a Meta's internal linter. I don't think there is a way to run it outside of Meta's infra.

codyschierbeck commented:
@mbasmanova Okay, I will try pushing these changes and see if there is still a Linter error afterwords.

rui-mo commented on velox/docs/functions/spark/regexp.rst:
```diff
@@ -34,3 +34,46 @@ See https://github.com/google/re2/wiki/Syntax for more information.
 
         SELECT rlike('1a 2b 14m', '\d+b'); -- true
+
+.. spark:function:: regex_replace(string, pattern, overwrite) -> varchar
```
Thanks for catching this! It's "regexp_replace" indeed.
https://github.com/apache/spark/blob/v3.2.2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L580
